usr/opt/OSC200/usr/share/man/man1/tabs.1
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: tabs.1,v $ $Revision: 3.2.5.3 $ (DEC) $Date: 1993/11/19 16:04:12 $
.\"
.\" 
.\" (c) Copyright 1991, OPEN SOFTWARE FOUNDATION, INC.
.\" ALL RIGHTS RESERVED
.\" 
.TH tabs 1
.SH NAME
\*Ltabs\*O \- Sets tab stops on terminals
.SH SYNOPSIS
.PP
.sS
\*Ltabs\*O
\*O[\*Vtab_spec\*O]
\*O[\*L+m\*Vnumber\*O]
\*O[\*L-T\*Vterminal\*O]
.sE
.PP
The \*Ltabs\*O command clears up to 20 previous tabs and sets up to
40 tabs on the terminal according to the supplied \*Vtab_spec\*O.
.SH FLAGS
.PP
If the same flag occurs more than once, only the last one takes effect.
.VL 4m
.LI "\*L-n\*O"
Specifies repetitive tab stops separated by a uniform number of column positions,
\*Ln\*O, where \*Ln\*O is a single-digit number. The default usage of \*Ltabs\*O
with no arguments is equivalent to tabs -8. If you use \*L-0\*O, the tab stops are
cleared and no new ones are set.
.LI "\*L-T\*Vterminal\*O"
Identifies the terminal so that \*Ltabs\*O can set tabs and
margins correctly.  The \*Vterminal\*O argument is one of the conventional
terminal names supported by your system.
.IP
If the \*Vterminal\*O you specify is not known to the system, \*Ltabs\*O
tries a general value that works for most terminals.
.IP
If you do not provide a \*L-T\*O flag, \*Ltabs\*O uses the
\*LTERM\*O shell variable.
.LI "\*L\(plm\*Vnumber\*O"
Moves all tabs to the right \*Vnumber\*O columns, and makes column
\*Vnumber\*L+1\*O the left margin.  (Note that columns = bytes.)
.sp
 If \*Lm\*O is given without a
value, 10 is assumed.  The leftmost margin on most terminals is
defined by \*Lm0\*O.
.LE
.SH DESCRIPTION
.iX "tabs command"
.iX "setting" "terminal tabs"
.iX "terminals" "setting tabs"
.iX "tabs" "setting"
.PP
The \*Vtab_spec\*O argument can be a flag representing one
of the pre-set tab specifications listed below.  It can also be an
alternative specification using either column numbers or the name of
an existing file, as described below.  If you do not
specify a \*Vtab_spec\*O, the default value is \*L-8\*O.
.PP
The available flags cover formats
required by most structured programming languages.
.PP
When you use the \*Ltabs\*O command, always consider the leftmost
column number to be 1, even if your terminal refers to it as 0 (zero).
Tab-stop position \*Ln\*O means that tabbing to position \*Ln\*O causes the
next character output to be in the \*Ln\*O+1th column position on that
line.
.SS Tab Specification Flags
.VL 4m
.LI "\*L-a\*O"
Sets the tabs to 1, 10, 16, 36, and 72.
.LI "\*L-a2\*O"
Sets the tabs to 1, 10, 16, 40, and 72.
.LI "\*L-c\*O"
Sets the tabs to 1, 8, 12, 16, 20, and 55 (COBOL normal format)
.LI "\*L-c2\*O"
Sets the tabs to 1, 6, 10, 14, and 49 (COBOL compact format, columns
1 to 6 omitted).  With this code, the first column position corresponds to
card column 7.  One space gets you to column 8, and a tab reaches column
12.  Files using this code must include the following format specification:
.iS
<:t-c2  m6  s66  d:>
\*O\c
.iE
.IP
See \*LFormat Specifications\*O later in this manpage.
.LI "\*L-c3\*O"
Sets the tabs to 1, 6, 10, 14, 18, 22, 26, 30, 34, 38, 42, 46,
50, 54, 58, 62, and 67 (COBOL compact format with more tabs
than \*L-c2\*O).  This is the recommended format for COBOL.
Files using this code must include the following format specification:
.iS
<:t-c3  m6  s66  d:>
\*O\c
.iE
.LI "\*L-f\*O"
Sets the tabs to 1, 7, 11, 15, 19, and 23 (FORTRAN).
.LI "\*L-p\*O"
Sets the tabs to 1, 5, 9, 13, 17, 21, 25, 29, 33, 37,
41, 45, 49, 53, 57, and 61 (PL/I).
.LI "\*L-s\*O"
Sets the tabs to 1, 10, and 55 (SNOBOL).
.LI "\*L-u\*O"
Sets the tabs to 1, 12, 20, and 44.
.LE
.PP
In addition to the preset formats, three other types
of \*Vtab_spec\*Os are available:
.VL 4m
.LI "\*L-\*Vnumber\*O"
Sets regularly repeating tabs at every \*Vnumber\*O column.
The standard tab setting is \*L-8\*O and is the setting required
for use with the \*Lnroff -h\*O flag.
Another special case is \*L-0\*O, which implies no tabs at all.
.LI "\*Vnumber\*O[,\*Vnumber\*O] \&.\&.\&."
Sets tabs at the named column numbers (a list in ascending order,
separated by commas).  You can specify up to 40 numbers.  If any number
except the first has a plus sign prefix, the prefixed number is
added to the previous number for the next setting.
Thus, the tab lists \*L1,10,20,30\*O and \*L1,10,+10,+10\*O
provide the same tab settings.
.LE
.SS Format Specifications
.IP
A format specification consists of a sequence of arguments separated by blanks
and enclosed in brackets and colons: \*L<: :>\*O.  Each argument consists
of a keyletter and an optional value which immediately follows it.
The following arguments can be used:
.VL 4m
.LI "\*Lt\*Vtabs\*O"
Specifies tab settings.  \*Vtabs\*O can be: a list of column numbers separated
by commas, indicating tab stops at the specified columns; a \*L-\*O (dash)
followed by a number \*Vn\*O, specifying tabs stops every \*Vn\*O columns;
or a \*L-\*O (dash) followed by a pre-set tab specification, for example
\*La2\*O (see \*LDESCRIPTION\*O).
.LI "\*Ls\*Vsize\*O"
Specifies a maximum line size, or length.  The \*Vsize\*O specification must be an integer.
(\*Vsize\*O is checked after tabs have been expanded but before the
\*Vmargin\*O is prepended.)
.LI "\*Lm\*Vmargin\*O"
Specifies a number of spaces to be inserted at the beginning of each line.
The \*Vmargin\*O specification must be an integer.
.LI "\*Ld\*O"
Indicates that the line containing the format specification is to be deleted
from the converted file.
.LI "\*Le\*O"
Indicates that the current format should prevail only until another format
specification is encountered in the file.
.LE
.PP
Default values of \*Lt-8\*O and \*Lm0\*O are assumed if \*Lt\*O and \*Lm\*O
arguments are not included in the specification; if \*Ls\*O is not included,
line size is not checked.  If the first line of a file does not contain a
format specification, these defaults are assumed for the entire file.
.PP
If a format specification can be disguised as a comment, for example
\*L* <:t5,10,15 s75 m5:> *\*O, you do not need to include the \*Ld\*O
keyletter.
.SH EXAMPLES
.PP
To set tabs every four spaces, enter:
.iS
tabs -4
\*O\c
.iE
.SH RELATED INFORMATION
.PP
Commands: \*Lnroff\*O(1), \*Ltset\*O(1).
.EQ
delim off
.EN


usr/opt/OSCX200/src/clients/xterm/tabs.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
/*
 *	$XConsortium: tabs.c,v 1.4 91/05/06 17:12:18 gildea Exp $
 */

/*
 * Copyright 1987 by Digital Equipment Corporation, Maynard, Massachusetts.
 *
 *                         All Rights Reserved
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation, and that the name of Digital Equipment
 * Corporation not be used in advertising or publicity pertaining to
 * distribution of the software without specific, written prior permission.
 *
 *
 * DIGITAL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
 * ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
 * DIGITAL BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
 * ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
 * WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
 * ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

/* tabs.c */

#include "ptyx.h"

/*
 * This file presumes 32bits/word.  This is somewhat of a crock, and should
 * be fixed sometime.
 */

/*
 * places tabstops at only every 8 columns
 */
TabReset(tabs)
Tabs	tabs;
{
	register int i;

	for (i=0; i<TAB_ARRAY_SIZE; ++i)
		tabs[i] = 0;

	for (i=0; i<MAX_TABS; i+=8)
		TabSet(tabs, i);
}	


/*
 * places a tabstop at col
 */
TabSet(tabs, col)
    Tabs	tabs;
    int		col;
{
	tabs[col >> 5] |= (1 << (col & 31));
}

/*
 * clears a tabstop at col
 */
TabClear(tabs, col)
    Tabs	tabs;
    int		col;
{
	tabs[col >> 5] &= ~(1 << (col & 31));
}

/*
 * returns the column of the next tabstop
 * (or MAX_TABS - 1 if there are no more).
 * A tabstop at col is ignored.
 */
TabNext (tabs, col)
    Tabs	tabs;
    int		col;
{
	extern XtermWidget term;
	register TScreen *screen = &term->screen;

	if(screen->curses && screen->do_wrap && (term->flags & WRAPAROUND)) {
		Index(screen, 1);
		col = screen->cur_col = screen->do_wrap = 0;
	}
	for (++col; col<MAX_TABS; ++col)
		if (tabs[col >> 5] & (1 << (col & 31)))
			return (col);

	return (MAX_TABS - 1);
}

/*
 * clears all tabs
 */
TabZonk (tabs)
Tabs	tabs;
{
	register int i;

	for (i=0; i<TAB_ARRAY_SIZE; ++i)
		tabs[i] = 0;
}
